File: code/ex186.java
class Solution {
    public void reverseWords(char[] s) {
        int n = s.length;
        // Reverse the char array
        reverse(s, 0, n - 1);
        
        // Reverse each word
        int slow = 0, fast = 0;
        while (fast < n) {
            if (s[fast] == ' ') {
                reverse(s, slow, fast - 1);
                slow = fast + 1;
            }
            fast++;
        }
        reverse(s, slow, n - 1);
    }
    
    private void reverse(char[] s, int lo, int hi) {
        while (lo < hi) {
            char temp = s[lo];
            s[lo++] = s[hi];
            s[hi--] = temp;
        }
    }
}

File: code/ex186.cpp
class Solution {
public:
    void reverseWords(vector<char>& s) {
        int n = s.size();
        
        // Reverse all chars
        int lo = 0, hi = n - 1;
        while (lo < hi)    swap(s[lo++], s[hi--]);
        
        // Reverse each single word
        int prev = 0;
        for (int i = 0; i < n; i++) {
            if (s[i] == ' ') {
                lo = prev;    hi = i - 1;
                while (lo < hi)    swap(s[lo++], s[hi--]);
                prev = i + 1;
            }
        }
        
        // Ending
        lo = prev;    hi = n - 1;
        while (lo < hi)    swap(s[lo++], s[hi--]);
    }
};